### PR TITLE
Update changelog draft

### PR DESCRIPTION
**Pull request by ahk**
Update the `changelog-draft` tool to support authoring `CHANGELOG` entries with Pull Request bodies.

Feedback still needed on the exact labeling/parsing strategy for the `CHANGELOG` entries content within Pull Request bodies.

### CODE REVIEW

File: src/get-changelog-pull-requests.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    return found;
+}
+
+function categorizePullRequests(pullRequests) {
+    const hasChangelog = []
+    const skipChangelog = [];
+    const needsChangelog = [];
+
+    for (const pr of pullRequests) {
```
Comment (asheemmamoowala) [inline src/get-changelog-pull-requests.js]: The way I read the labels is to indicate whether the PR should or should not be included in the CHANGELOG. Where the label doesn't exist, it should default to inclusion in the draft (maybe with an 🤷‍♀ , to indicate lack of label)

A PR with the `needs changelog` label may or may not have a useful PR title or changelog suggestion in the body.

**Comment by ahk** [comment]
Having thought about what we ended up with, I'd like to recommend two things:

1. Instead of hand-typing the categories (and then parsing them out), we use the existing labels on a PR. We currently have `bug`, `breaking change`, `feature`, `docs`, `performance`, `workflow`, and `testing` as PR labels. These would be used to categorize the list of entries that are pulled into the draft. We wouldn't need `needs changelog`, anything without one of these labels or `skip changelog` will be included in the `other` category. Anything marked `skip changelog` will not be included in the draft.

2. A checkbox will be added to the PR template. It will be named "Includes changelog entry in PR body formatted as `changelog: ...`". Any text following `changelog:` until the end of the PR body will be parsed and used as the `CHANGELOG` entry instead of the title in this case.

**Comment by asheemmamoowala** [comment]
> Instead of hand-typing the categories (and then parsing them out), we use the existing labels on a PR. We currently have bug, breaking change, feature, docs, performance, workflow, and testing as PR labels. 

Doesn't the tool already do this by looking at labels on the referenced issues? It is not common practice to put labels on the PRs. 

> A checkbox will be added to the PR template. It will be named "Includes changelog entry in PR body formatted as changelog: ...". Any text following changelog: until the end of the PR body will be parsed and used as the CHANGELOG entry instead of the title in this case.

I think it would be better for the check box to just prompt for the changelog entry with some sort of start/end tags within which the changelog text is placed. This would simplify the instrcutions in the template and the parsing. Something like:

```
- [ ] Changelog draft: <changelog>Propose a changelog entry here</changelog>
```
- [ ] Changelog draft: <changelog>Propose a changelog entry here</changelog>

**Comment by ahk** [comment]
> I think it would be better for the check box to just prompt for the changelog entry with some sort of start/end tags within which the changelog text is placed.

Works for me.

> Doesn't the tool already do this by looking at labels on the referenced issues? It is not common practice to put labels on the PRs.

One of the things from the original proposal thread, and one of the main things that this PR does, is remove the logic that parsed linked issues and commits. This is because it was buggy (parsing PR body text and commits) and overly broad (any issue referenced in a PR was included, and attempts were made to categorize based on potentially multiple linked issues).

In my opinion the more text we include that may or may not be correct, but wasn't explicitly hand edited or tagged, the less useful a tool like this is. These labels exist for PRs already, and exactly match what I found in the original source for this tool, I guess we added them at some point but decided not to use them?

It is not currently the practice to hand tag PRs, but as I understood and summarized the original proposal that's something we meant to change. If we are doing that, then they are either parsed from text or found from the list of github labels.

If this isn't desired then we should I guess revert to the previous way this tool worked, with the slight addition of also including the PR title or optionally the parsed PR body if a `<changelog>entry text</changelog>` is found.

**Comment by asheemmamoowala** [comment]
@ahk You're right, the solution to tag PRs to aid in the changelog curation is helpful. Lets move forward with this PR as is and start incorporating it in our workflow.

File: src/render.js [hunk 2]
```diff
@@ -1,35 +1,49 @@
...
-exports.commitSummary = function commitSummary(commit) {
-    const {issues, pr} = commit;
-    const title = pr ? formatIssue(pr) : `${commit.title} ([${commit.sha.slice(0, 7)}](${commit.url}))`;
+    const prRef = ` ([#${pr.number}](${pr.html_url}))`;
 
     let hattip = '';
-    if (pr && pr.user) {
-        hattip = ` (h/t [${pr.user.login}](${pr.user.url}))`;
+    if (pr.head.user.login !== pr.base.user.login) {
+        hattip += ` (h/t [${pr.user.login}](${pr.user.url}))`;
```
Comment (kkaefer) [inline src/render.js]: This links to `api.github.com` but should link to the HTML profile on `github.com`

**Comment by kkaefer** [comment]
I used this to get started with the release changelog for GL JS 1.4.0. It worked well, except for the issue reported above. However, most of the issues were unclassified, likely because we didn't have tags applied to them.

**Comment by zugaldia** [comment]
fyi: @tobrun @julianrex @tmpsantos  @chloekraw -- we were recently discussing CHANGELOG entries in the context of the new modular SDK repos, I'd love to see if we can take advantage of this tooling as well.

**Comment by julianrex** [comment]
@ahk does this allow for multiple entries in a PR? I'm thinking of the case where we update to a new version of a dependency; ideally we'd like to be able to expose particular change-log entries from the dependency in our SDK's change-log.

**Comment by ahk** [comment]
> @ahk does this allow for multiple entries in a PR?

Currently it does not. We had some long conversations about this and I advocated for supporting multiple entries per PR, but several people stated they didn't think we needed that historically, and that it was too much complexity.

Right now this tool assumes only one entry, and we get its categories from github PR labels.

If we did support multiple entries we'd need to agree on a simple formatting rule for the category tags like `<bug>Here is some changelog text about this bug</bug><feature>Here is something about this separate feature</feature>` or something similarly easy to parse. And that should be located somewhere in the PR body.

Would love to hear your thoughts on this and I'm happy to extend this tool. Any chance you could point me to some notes on how gl-native builds their changelogs?

**Comment by julianrex** [comment]
cc @captainbarbosa

**Comment by julianrex** [comment]
> If we did support multiple entries we'd need to agree on a simple formatting rule for the category tags like <bug>Here is some changelog text about this bug</bug><feature>Here is something about this separate feature</feature> or something similarly easy to parse. And that should be located somewhere in the PR body.

This sounds like an addition, rather than changing what you already have, so that's a 👍 from me. Format wise - I'd pick whatever's easiest for the dev to write & then the tool to parse. Perhaps we could reference the original PR, and the process is recursive? So something like
```
[label-name](link to original PR)
```
would pull in the change log entries from that PR into the parent. 

This doesn't allow for modifying the entry, but perhaps not a huge issue.

Thoughts @tobrun @captainbarbosa?

> Any chance you could point me to some notes on how gl-native builds their changelogs?

For iOS/macOS these are hand-crafted, according to these notes (which have aged): https://github.com/mapbox/mapbox-gl-native/wiki/Release-notes-style-guide. Though we do want to automate this process (ideally with this tool!) and change the headings along similar lines as this PR.

**Comment by ahk** [comment]
> This doesn't allow for modifying the entry, but perhaps not a huge issue.

If it pulls the text from the PR you can always edit the content there before running this tool right? It was something we considered when we decided to move to PR text. Also you are free to edit the text made by this tool before committing the results since it just builds a draft for you.

Adding the ability to recursively pull from referenced PRs is fine by me, and yes is an addition to our (very simple) process. One thing we remove with these changes is the reliance on walking through commit messages and referenced issues, because it was too much noise. My personal opinion is adding back in reference walking for PRs would be a good compromise, as they are single points for the change to be introduced, are directly reviewed, and can be updated without requiring codebase changes.

@julianrex This PR has been sitting for a while and just needs a small bugfix from me for GL-JS to start using it officially, do you mind if I merge this and we can keep this discussion going when native is ready to add some features?

I look forward to working more with you all to improve/streamline our changelog process and policy. Changelogs are important!

**Pull request merged**